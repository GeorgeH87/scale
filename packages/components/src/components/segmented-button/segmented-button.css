/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  /* not selected */
  --background-color-selected: var(--telekom-color-ui-base);
  --button-radius: 6px;
  --padding-top-bottom: var(--telekom-spacing-unit-x1);
  --padding-left-right: var(--telekom-spacing-unit-x2);
  --font-family: var(--telekom-typography-font-family-sans);
  --font-weight: var(--telekom-typography-font-weight-bold);
  --font-size: var(--telekom-typography-font-size-caption);
  --label-disabled: var(--telekom-color-text-and-icon-disabled);
  --transition: all var(--telekom-motion-duration-transition)
    var(--telekom-motion-easing-standard);

  --color-selected: var(--telekom-color-text-and-icon-primary-standard);

  /* large styles */
  --font-size-large: var(--telekom-typography-font-size-body);
  --large-height: var(--telekom-spacing-unit-x6);

  /* xl styles */
  --font-size-xl: var(--telekom-typography-font-size-body);
  --xl-height: var(--telekom-spacing-unit-x9);

  /* hover styles */
  --background-color-hover: var(--telekom-color-ui-state-fill-hovered);

  /* active styles */
  --background-color-active: var(--telekom-color-ui-state-fill-pressed);
  --box-shadow-focus: 0 0 0 var(--telekom-line-weight-highlight)
    var(--telekom-color-functional-focus);

  display: flex;
  align-items: center;
  justify-content: center;
  flex: 1;
}

.segmented-button {
  box-sizing: border-box;
  position: relative;
  border: 0;
  height: 100%;
  width: 100%;
  border-radius: var(--button-radius);
  padding-top: var(--padding-top-bottom);
  padding-bottom: var(--padding-top-bottom);
  transition: var(--transition);
  background-color: transparent;
  padding: 0;
}

.segmented-button--mask {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  height: 100%;
  width: 100%;
  border-radius: var(--button-radius);
  background-color: transparent;  
  font-family: var(--font-family);
  font-size: var(--font-size);
  font-weight: var(--font-weight);
  transition: var(--transition);
  white-space: nowrap;
  color: var(--telekom-color-text-and-icon-standard);
}

button.segmented-button--icon-only.segmented-button--small .segmented-button--mask {
  padding-left: var(--telekom-spacing-unit-x4);
  padding-right: var(--telekom-spacing-unit-x4);
}

button.segmented-button--small .segmented-button--mask   {
  padding-left: var(--telekom-spacing-unit-x2);
  padding-right: var(--telekom-spacing-unit-x3);
}

.segmented-button--large {
  font-size: var(--font-size-large);
  height: 36px;
}

button.segmented-button--large .segmented-button--mask {
  padding-left: var(--telekom-spacing-unit-x3);
  padding-right: var(--telekom-spacing-unit-x4);
}

.segmented-button--xl {
  font-size: var(--font-size-large);
  height: 40px;
}
button.segmented-button--xl .segmented-button--mask {
  padding-left: 16px;
  padding-right: 20px;
}

.segmented-button--selected {
  background-color: var(--background-color-selected);
  box-shadow: var(--telekom-shadow-raised-standard);
  color: var(--color-selected);
}

.segmented-button--disabled.segmented-button--selected .segmented-button--mask {
  color: var(--label-disabled);
}

.segmented-button--disabled:not(.segmented-button--selected) .segmented-button--mask{
  color: var(--label-disabled);
  background-color: transparent;
}

.segmented-button--left-sibling-selected {
  border-radius: 0 var(--button-radius) var(--button-radius) 0;
  margin-left: 0;
  clip-path: inset(-44px -44px -44px 0px);
  backface-visibility: hidden;
  overflow: hidden;
}

.segmented-button--right-sibling-selected {
  border-radius: var(--button-radius) 0 0 var(--button-radius);
  clip-path: inset(-44px -0px -44px -44px);
  margin-right: 0;
  backface-visibility: hidden;
  overflow: hidden;
  position: relative;
}

.segmented-button--left-right-sibling-selected {
  clip-path: inset(-44px 0px -44px -0px);
  border-radius: 0;
  margin-left: 0;
  margin-right: 0;
  backface-visibility: hidden;
}

.segmented-button:not(.segmented-button--disabled):hover .segmented-button--mask{
  background-color: var(--background-color-hover);
}

.segmented-button:not(.segmented-button--disabled):active .segmented-button--mask{
  background-color: var(--background-color-active);
}

.segmented-button:focus {
  box-shadow: var(--telekom-shadow-raised-standard);
  outline: var(--telekom-spacing-unit-x05) solid var(--telekom-color-functional-focus );
  outline-offset: var(--telekom-spacing-unit-x025);
  z-index: 40;
  position: relative;
}

.segmented-button--left-right-sibling-selected:focus,
.segmented-button--right-sibling-selected:focus,
.segmented-button--left-sibling-selected:focus {
  clip-path: none;
}

.segmented-button:not(.segmented-button--icon-only) scale-icon-action-success,
.segmented-button scale-icon-action-close {
  margin-right: var(--telekom-spacing-unit-x1);
}

.segmented-button:not(.segmented-button--selected) scale-icon-action-success {
  visibility: hidden !important;
} 

.segmented-button scale-icon-action-success {
  vertical-align: middle;
}

.segmented-button--selected:not(.segmented-button--icon-only) .icon-container {
  display: none !important;
}

.segmented-button .success-icon-container {
  display: none; 
}

.segmented-button--selected .success-icon-container {
  display: flex;
  justify-content: center;
  align-items: center;
}

.icon-container {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.segmented-button--selected .icon-container {
  color: var(--color-selected);
}
