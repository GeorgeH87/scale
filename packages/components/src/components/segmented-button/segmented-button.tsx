/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Component,
  Prop,
  h,
  Host,
  Element,
  State,
  Listen,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';
import classNames from 'classnames';
import { emitEvent } from '../../utils/utils';

interface SegmentStatus {
  id: string;
  selected: boolean;
}

const CHECKMARK_WIDTH_SMALL = 14;
const CHECKMARK_WIDTH_MEDIUM = 18 + 12;
const CHECKMARK_WIDTH_LARGE = 20 + 18;

@Component({
  tag: 'scale-segmented-button',
  styleUrl: 'segmented-button.css',
  shadow: true,
})
export class SegmentedButton {
  /** segment position within button */
  position = 0;

  slottedSegments = 0;

  @Element() hostElement: HTMLElement;
  /** state */
  @State() status: SegmentStatus[] = [];
  /** (optional) The size of the button */
  @Prop() size?: 'small' | 'medium' | 'large' = 'small';
  /** (optional) Allow more than one button to be selected */
  @Prop() multiSelect: boolean = false;
  /** (optional) If `true`, the button is disabled */
  @Prop({ reflect: true }) disabled?: boolean = false;
  /** (optional) If `true`, expand to container width */
  @Prop() fullWidth?: boolean = false;
  /** (optional) If `true`, show error message */
  @Prop() invalid?: boolean = false;
  /** (optional) If `true`, show error message */
  @Prop() helperText?: string = 'Please select an option';
  /** (optional) Button label */
  @Prop() label?: string;
  /** (optional) icon only */
  @Prop() iconOnly?: boolean = false;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;
  /** (optional) aria-label attribute needed for icon-only buttons */
  @Prop()
  ariaLabelTranslation = `segment button with $slottedSegments elements`;
  @Prop({ mutable: true })
  longestButtonWidth: string;
  /** Emitted when button is clicked */
  @Event({ eventName: 'scale-change' }) scaleChange: EventEmitter;
  /** @deprecated in v3 in favor of kebab-case event names */
  @Event({ eventName: 'scaleChange' }) scaleChangeLegacy: EventEmitter;

  container: HTMLElement;
  showHelperText = false;

  @Listen('scaleClick')
  scaleClickHandler(ev: { detail: { id: string; selected: boolean } }) {
    let tempState: SegmentStatus[];
    if (!this.multiSelect) {
      if (!ev.detail.selected) {
        tempState = this.status.map((obj) =>
          ev.detail.id === obj.id ? ev.detail : { ...obj }
        );
        /* clicked button has now selected state */
      } else {
        tempState = this.status.map((obj) =>
          ev.detail.id === obj.id ? ev.detail : { ...obj, selected: false }
        );
      }
    } else {
      tempState = this.status.map((obj) =>
        ev.detail.id === obj.id ? ev.detail : { ...obj }
      );
    }
    this.setState(tempState);
  }

  @Watch('disabled')
  @Watch('size')
  handlePropsChange() {
    this.propagatePropsToChildren();
  }

  /**
   * Keep props, needed in children buttons, in sync
   */
  propagatePropsToChildren() {
    this.getAllSegmentedButtons().forEach((el) => {
      el.setAttribute('size', this.size);
      if (this.disabled) {
        el.setAttribute('disabled', '');
      }
      if (this.multiSelect) {
        el.setAttribute('multi-select', '');
      }
    });
  }

  connectedCallback() {
    this.propagatePropsToChildren();
  }

  componentDidLoad() {
    const tempState: SegmentStatus[] = [];
    const segmentedButtons = this.getAllSegmentedButtons();
    this.slottedSegments = segmentedButtons.length;
    segmentedButtons.forEach((segment) => {
      this.position++;
      tempState.push({
        id: segment.getAttribute('segment-id'),
        selected: segment.hasAttribute('selected'),
      });
      segment.setAttribute('position', this.position.toString());
      segment.setAttribute(
        'aria-description-translation',
        '$position $selected'
      );
    });
    this.hostElement.style.setProperty(
      '--colNum',
      this.slottedSegments.toString()
    );
    this.position = 0;
    this.status = tempState;
    this.setState(tempState);
  }

  componentWillUpdate() {
    this.showHelperText = false;
    if (
      this.invalid &&
      this.status.filter((e) => e.selected === true).length <= 0
    ) {
      this.showHelperText = true;
    }
  }

  getAdjacentSiblings = (tempState, i) => {
    let adjacentSiblings = '';
    if (i !== 0 && tempState[i].selected && tempState[i - 1].selected) {
      adjacentSiblings = 'left';
    }
    if (
      i !== tempState.length - 1 &&
      tempState[i].selected &&
      tempState[i + 1].selected
    ) {
      adjacentSiblings = `${
        adjacentSiblings ? adjacentSiblings + ' right' : 'right'
      }`;
    }
    return adjacentSiblings;
  };

  // all segmented buttons should have the same width, based on the largest one
  getLongestButtonWidth() {
    let tempWidth = 0;
    Array.from(this.hostElement.children).forEach((child) => {
      const selected = child.hasAttribute('selected');
      const iconOnly = child.hasAttribute('icon-only');
      const checkmark =
        this.size === 'small'
          ? CHECKMARK_WIDTH_SMALL
          : this.size === 'medium'
          ? CHECKMARK_WIDTH_MEDIUM
          : CHECKMARK_WIDTH_LARGE;
      if (selected || iconOnly) {
        tempWidth =
          child.getBoundingClientRect().width > tempWidth
            ? child.getBoundingClientRect().width
            : tempWidth;
      } else {
        tempWidth =
          child.getBoundingClientRect().width + checkmark > tempWidth
            ? child.getBoundingClientRect().width + checkmark
            : tempWidth;
      }
    });
    return tempWidth;
  }

  setState(tempState: SegmentStatus[]) {
    const segmentedButtons = Array.from(
      this.hostElement.querySelectorAll('scale-segment')
    );
    segmentedButtons.forEach((segmentedButton, i) => {
      segmentedButton.setAttribute(
        'adjacent-siblings',
        this.getAdjacentSiblings(tempState, i)
      );
      segmentedButton.setAttribute(
        'selected',
        tempState[i].selected ? 'true' : 'false'
      );
    });
    this.status = tempState;
    emitEvent(this, 'scaleChange', this.status);
  }

  getAllSegmentedButtons() {
    return Array.from(this.hostElement.querySelectorAll('scale-segment'));
  }

  getAriaLabelTranslation() {
    const filledText = this.ariaLabelTranslation.replace(
      /\$slottedSegments/g,
      `${this.slottedSegments}`
    );
    return filledText;
  }

  render() {
    return (
      <Host>
        <fieldset
          class="segmented-button--fieldset"
          aria-label={
            this.showHelperText && this.helperText ? this.helperText : null
          }
        >
          {this.label && (
            <legend class="segmented-button--label"> {this.label} </legend>
          )}
          <ul
            class={this.getCssClassMap()}
            part={this.getBasePartMap()}
            aria-label={this.getAriaLabelTranslation()}
            role="listbox"
            ref={(el) => (this.container = el as HTMLUListElement)}
          >
            <slot />
          </ul>
          {this.showHelperText && (
            <scale-helper-text
              class="segmented-button--helper-text"
              helperText={this.helperText}
              variant={'danger'}
            ></scale-helper-text>
          )}
        </fieldset>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const prefix = mode === 'basePart' ? '' : 'segmented-button--';

    return classNames(
      'segmented-button',
      this.size && `${prefix}${this.size}`,
      this.fullWidth && `${prefix}full-width`
    );
  }
}
