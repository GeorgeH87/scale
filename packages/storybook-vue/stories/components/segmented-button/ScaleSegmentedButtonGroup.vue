<template>
  <scale-segmented-button-group
    :size="size"
    :disabled="disabled"
    :multi-select="multiSelect"
    :required="required"
    :label="label"
    :helper-text="helperText"
    :styles="styles"
    @scaleChange="scaleChange"
  >
    <slot></slot>
  </scale-segmented-button-group>
</template>

<script>
import { action } from "@storybook/addon-actions";
export default {
  props: {
    size: { type: String, default: 'regular' },
    disabled: { type: Boolean, default: false },
    multiSelect: { type: Boolean, default: false },
    required: {type: Boolean, default: false},
    label: {type: String},
    helperText: {type: String},
    styles: { type: String },
  },
  methods: {
    scaleChange($event) {
      action("scaleChange");
      this.$emit("scaleChange", $event);
    },
    'scale-change'($event) {
      action("scale-change");
      this.$emit("scale-change", $event);
    },
  }
};
</script>