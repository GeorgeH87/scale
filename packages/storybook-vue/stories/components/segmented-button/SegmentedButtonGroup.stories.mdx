import { Meta, Story, ArgsTable, Canvas } from '@storybook/addon-docs';
import ScaleSegmentedButtonGroup from './ScaleSegmentedButtonGroup.vue';

<Meta
  title="Beta Components/Segmented Button"
  component={ScaleSegmentedButtonGroup}
  argTypes={{
    size: {
      control: { type: 'select', options: ['small', 'medium', 'large'] },
    },
  }}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleSegmentedButtonGroup },
  props: {
    label: String,
    icon: String,
    ...ScaleSegmentedButtonGroup.props,
  },
  template: `
    <scale-segmented-button-group
      :size="size"
      :disabled="disabled"
      :multi-select="multiSelect"
      :styles="styles"
      :invalid="invalid"
      :helper-text="helperText"
      :label="label"
    >
      <scale-segmented-button selected>Apple</scale-segmented-button>
      <scale-segmented-button>One+</scale-segmented-button>
      <scale-segmented-button>Samsung</scale-segmented-button>
      <scale-segmented-button>Huawei</scale-segmented-button>
    </scale-segmented-button-group>
  `,
});

export const MultiSelectTemplate = (args, { argTypes }) => ({
  components: { ScaleSegmentedButtonGroup },
  props: {
    label: String,
    icon: String,
    ...ScaleSegmentedButtonGroup.props,
  },
  template: `
    <scale-segmented-button-group
      :size="size"
      :disabled="disabled"
      :multi-select="multiSelect"
      :styles="styles"
    >
      <scale-segmented-button selected>Apple</scale-segmented-button>
      <scale-segmented-button selected>One+</scale-segmented-button>
      <scale-segmented-button>Samsung</scale-segmented-button>
      <scale-segmented-button>Huawei</scale-segmented-button>
    </scale-segmented-button-group>
  `,
});

export const IconOnlyTemplate = (args, { argTypes }) => ({
  components: { ScaleSegmentedButtonGroup },
  props: {
    label: String,
    icon: String,
    ...ScaleSegmentedButtonGroup.props,
  },
  template: `
    <scale-segmented-button-group
      :size="size"
      :disabled="disabled"
      :multi-select="multiSelect"
      :styles="styles"
    >
      <scale-segmented-button selected><scale-icon-weather-cloudy slot="segmented-button-icon"></scale-icon-weather-cloudy></scale-segmented-button>
      <scale-segmented-button><scale-icon-weather-rain slot="segmented-button-icon"></scale-icon-weather-rain></scale-segmented-button>
      <scale-segmented-button><scale-icon-weather-sunny slot="segmented-button-icon"></scale-icon-weather-sunny></scale-segmented-button>
      <scale-segmented-button><scale-icon-weather-heavy-snow slot="segmented-button-icon"></scale-icon-weather-heavy-snow></scale-segmented-button>
    </scale-segmented-button-group>
  `,
});

export const IconTextTemplate = (args, { argTypes }) => ({
  components: { ScaleSegmentedButtonGroup },
  props: {
    label: String,
    icon: String,
    ...ScaleSegmentedButtonGroup.props,
  },
  template: `
    <scale-segmented-button-group
      :size="size"
      :disabled="disabled"
      :multi-select="multiSelect"
      :styles="styles"
    >
      <scale-segmented-button selected><scale-icon-device-device-phone slot="segmented-button-icon"></scale-icon-device-device-phone>Apple</scale-segmented-button>
      <scale-segmented-button><scale-icon-device-device-phone slot="segmented-button-icon"></scale-icon-device-device-phone>One+</scale-segmented-button>
      <scale-segmented-button><scale-icon-device-device-phone slot="segmented-button-icon"></scale-icon-device-device-phone>Samsung</scale-segmented-button>
      <scale-segmented-button><scale-icon-device-device-phone slot="segmented-button-icon"></scale-icon-device-device-phone>Huawei</scale-segmented-button>
    </scale-segmented-button-group>
  `,
});

export const DisabledGroupTemplate = (args, { argTypes }) => ({
  components: { ScaleSegmentedButtonGroup },
  props: {
    label: String,
    icon: String,
    disabled: 'true',
    ...ScaleSegmentedButtonGroup.props,
  },
  template: `
    <scale-segmented-button-group
      :disabled="disabled"
      :size="size"
      :multi-select="multiSelect"
      :styles="styles"
      :invalid="invalid"
      :helper-text="helperText"
    >
      <scale-segmented-button>Apple</scale-segmented-button>
      <scale-segmented-button selected>One+</scale-segmented-button>
      <scale-segmented-button>Samsung</scale-segmented-button>
      <scale-segmented-button>Huawei</scale-segmented-button>
    </scale-segmented-button-group>
  `,
});

export const DisabledTemplate = (args, { argTypes }) => ({
  components: { ScaleSegmentedButtonGroup },
  props: {
    label: String,
    icon: String,
    disabled: true,
    ...ScaleSegmentedButtonGroup.props,
  },
  template: `
    <scale-segmented-button-group
      :size="size"
      :disabled="disabled"
      :multi-select="multiSelect"
      :styles="styles"
      :invalid="invalid"
      :helper-text="helperText"
    >
      <scale-segmented-button disabled>Apple</scale-segmented-button>
      <scale-segmented-button>One+</scale-segmented-button>
      <scale-segmented-button>Samsung</scale-segmented-button>
      <scale-segmented-button>Huawei</scale-segmented-button>
    </scale-segmented-button-group>
  `,
});

export const InvalidTemplate = (args, { argTypes }) => ({
  components: { ScaleSegmentedButtonGroup },
  props: {
    label: String,
    icon: String,
    invalid: Boolean,
    ...ScaleSegmentedButtonGroup.props,
  },
  template: `
    <scale-segmented-button-group
      :invalid="invalid"
      :helper-text="helperText"
      :label="label"
    >
      <scale-segmented-button>Apple</scale-segmented-button>
      <scale-segmented-button>One+</scale-segmented-button>
      <scale-segmented-button>Samsung</scale-segmented-button>
      <scale-segmented-button>Huawei</scale-segmented-button>
    </scale-segmented-button-group>
  `,
});

## Standard

<Canvas withSource="none">
  <Story name="Standard" args={{ label: 'Group Label' }}>
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

```html
<scale-segmented-button-group>
  <scale-segmented-button selected>Apple</scale-segmented-button>
  <scale-segmented-button>One+</scale-segmented-button>
  <scale-segmented-button>Samsung</scale-segmented-button>
  <scale-segmented-button>Huawei</scale-segmented-button>
</scale-segmented-button-group>
```

## Multi Select

<Canvas withSource="none">
  <Story name="MultiSelect" args={{ multiSelect: true, label: 'Group Label' }}>
    {MultiSelectTemplate.bind({})}
  </Story>
</Canvas>

```html
<scale-segmented-button-group multi-select disabled>
  <scale-segmented-button>Apple</scale-segmented-button>
  <scale-segmented-button>One+</scale-segmented-button>
  <scale-segmented-button selected>Samsung</scale-segmented-button>
  <scale-segmented-button>Huawei</scale-segmented-button>
</scale-segmented-button-group>
```

## Disabled

### Disabled Button

<Canvas withSource="none">
  <Story name="Disabled Button" args={{ label: 'Group Label' }}>
    {DisabledTemplate.bind({})}
  </Story>
</Canvas>

```html
<scale-segmented-button-group>
  <scale-segmented-button disabled>Apple</scale-segmented-button>
  <scale-segmented-button>One+</scale-segmented-button>
  <scale-segmented-button>Samsung</scale-segmented-button>
  <scale-segmented-button>Huawei</scale-segmented-button>
</scale-segmented-button-group>
```

### Disabled Button Group

<Canvas withSource="none">
  <Story
    name="Disabled Button Group"
    args={{ label: 'Group Label', disabled: true }}
  >
    {DisabledGroupTemplate.bind({})}
  </Story>
</Canvas>

```html
<scale-segmented-button-group disabled>
  <scale-segmented-button>Apple</scale-segmented-button>
  <scale-segmented-button selected>One+</scale-segmented-button>
  <scale-segmented-button>Samsung</scale-segmented-button>
  <scale-segmented-button>Huawei</scale-segmented-button>
</scale-segmented-button-group>
```

## Icon only

<Canvas withSource="none">
  <Story name="Icon only" args={{ label: 'Group Label' }}>
    {IconOnlyTemplate.bind({})}
  </Story>
</Canvas>

```html
<scale-segmented-button-group>
  <scale-segmented-button selected
    ><scale-icon-weather-cloudy
      slot="segmented-button-icon"
    ></scale-icon-weather-cloudy
  ></scale-segmented-button>
  <scale-segmented-button
    ><scale-icon-weather-rain
      slot="segmented-button-icon"
    ></scale-icon-weather-rain
  ></scale-segmented-button>
  <scale-segmented-button
    ><scale-icon-weather-sunny
      slot="segmented-button-icon"
    ></scale-icon-weather-sunny
  ></scale-segmented-button>
  <scale-segmented-button
    ><scale-icon-weahter-heavy-snow
      slot="segmented-button-icon"
    ></scale-icon-weahter-heavy-snow
  ></scale-segmented-button>
</scale-segmented-button-group>
```

## Icon and Text

<Canvas withSource="none">
  <Story name="Icon and Text" args={{ label: 'Group Label' }}>
    {IconTextTemplate.bind({})}
  </Story>
</Canvas>

```html
<scale-segmented-button-group>
  <scale-segmented-button
    ><scale-icon-device-device-phone
      slot="segmented-button-icon"
    ></scale-icon-device-device-phone
    >Apple</scale-segmented-button
  >
  <scale-segmented-button
    ><scale-icon-device-device-phone
      slot="segmented-button-icon"
    ></scale-icon-device-device-phone
    >One+</scale-segmented-button
  >
  <scale-segmented-button selected
    ><scale-icon-device-device-phone
      slot="segmented-button-icon"
    ></scale-icon-device-device-phone
    >Samsung</scale-segmented-button
  >
  <scale-segmented-button selected
    ><scale-icon-device-device-phone
      slot="segmented-button-icon"
    ></scale-icon-device-device-phone
    >Huawei</scale-segmented-button
  >
</scale-segmented-button-group>
```

## Invalid

<Canvas withSource="none">
  <Story
    name="invalid"
    args={{
      invalid: true,
    }}
  >
    {InvalidTemplate.bind({})}
  </Story>
</Canvas>

```html
<scale-segmented-button-group invalid>
  <scale-segmented-button
    ><scale-icon-device-device-phone
      slot="segmented-button-icon"
    ></scale-icon-device-device-phone
    >Apple</scale-segmented-button
  >
  <scale-segmented-button
    ><scale-icon-device-device-phone
      slot="segmented-button-icon"
    ></scale-icon-device-device-phone
    >One+</scale-segmented-button
  >
  <scale-segmented-button
    ><scale-icon-device-device-phone
      slot="segmented-button-icon"
    ></scale-icon-device-device-phone
    >Samsung</scale-segmented-button
  >
  <scale-segmented-button
    ><scale-icon-device-device-phone
      slot="segmented-button-icon"
    ></scale-icon-device-device-phone
    >Huawei</scale-segmented-button
  >
</scale-segmented-button-group>
```