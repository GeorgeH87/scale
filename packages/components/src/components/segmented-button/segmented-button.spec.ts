import { newSpecPage } from '@stencil/core/testing';
import { SegmentedButton } from './segmented-button';

describe('SegmentedButton', () => {
  it('should match standard snapshot', async () => {
    const page = await newSpecPage({
      components: [SegmentedButton],
      html: `
        <scale-segmented-button>Label</scale-segmented-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });
  it('should include right properties', async () => {
    const page = await newSpecPage({
      components: [SegmentedButton],
      html: `
            <scale-segmented-button>Label</scale-segmented-button>`,
    });
    page.root.disabled = true;
    page.root.size = 'medium',
    page.root.selected = true;
    page.root.hasIcon = true;
    page.root.ariaLangSelected = 'ariaLangSelected';
    page.root.ariaLangDeselected = 'ariaLangDeselected';
    page.root.ariaDescriptionTranslation = 'ariaDescriptionTranslation';
    await page.waitForChanges();
    expect(page.rootInstance.disabled).toBe(true);
    expect(page.rootInstance.size).toBe('medium');
    expect(page.rootInstance.selected).toBe(true);
    expect(page.rootInstance.hasIcon).toBe(true);
    expect(page.rootInstance.ariaLangSelected).toBe('ariaLangSelected');
    expect(page.rootInstance.ariaLangDeselected).toBe('ariaLangDeselected');
    expect(page.rootInstance.ariaDescriptionTranslation).toBe('ariaDescriptionTranslation');
});
it('should include iconOnly', async () => {
    const page = await newSpecPage({
      components: [SegmentedButton],
      html: `
            <scale-segmented-button>Label</scale-segmented-button>`,
    });
    page.root.iconOnly = true;
   await page.waitForChanges();
    expect(page.rootInstance.iconOnly).toBe(true);
});
it('should include textOnly', async () => {
    const page = await newSpecPage({
      components: [SegmentedButton],
      html: `
            <scale-segmented-button>Label</scale-segmented-button>`,
    });
    page.root.textOnly = true;
   await page.waitForChanges();
    expect(page.rootInstance.textOnly).toBe(true);
});
});
