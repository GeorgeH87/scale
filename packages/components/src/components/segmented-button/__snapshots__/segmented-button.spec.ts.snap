// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedButton should match standard snapshot 1`] = `
<scale-segmented-button segmented-button-id="segmented-button-0">
  <mock:shadow-root>
    <button aria-description="deselected" class="segmented-button segmented-button--small" id="segmented-button-0" part="segmented-button small" type="button">
      <div class="segmented-button--mask">
        <div class="success-icon-container">
          <scale-icon-action-success accessibility-title="success" class="scale-icon-action-success" selected="" size="14"></scale-icon-action-success>
        </div>
        <div class="icon-container">
          <slot name="segmented-button-icon"></slot>
        </div>
        <slot></slot>
      </div>
    </button>
  </mock:shadow-root>
  Label
</scale-segmented-button>
`;
